teps/0141-v1-migration-for-tekton-results.md

```diff
@@ -0,0 +1,216 @@
+---
+status: proposed
+title: Migration of Tekton Results to V1 APIs of Pipelines
+creation-date: '2023-08-14'
+last-updated: '2023-08-14'
+authors:
+- '@khrm'
+- '@adambkaplan'
+---
+
+# TEP-0141: Migration of Tekton Results to V1 APIs of Pipelines
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Do Nothing and just shift the watch object to v1](#do-nothing-and-just-shift-the-watch-object-to-v1)
+    - [Cons](#cons)
+  - [A conversion function to render all stored objects in v1](#a-conversion-function-to-render-all-stored-objects-in-v1)
+    - [Cons](#cons-1)
+  - [Convert v1beta1 object for the first time they are accessed](#convert-v1beta1-object-for-the-first-time-they-are-accessed)
+    - [Cons](#cons-2)
+    - [A job to convert all v1beta1 objects](#a-job-to-convert-all-v1beta1-objects)
+  - [Notes and Caveats](#notes-and-caveats)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [Conformance](#conformance)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Implementation Plan](#implementation-plan)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+We want to add support for v1 APIs in Tekton results in a backward-compatible way. At present, we watch v1beta1 objects.
+
+## Motivation
+
+v1beta1 APIs are being deprecated and we want to shift to v1 APIs.
+
+### Goals
+
+* Watching v1 objects.
+* Storing v1 objects.
+* Handling existing v1beta1 objects in the database.
+
+### Non-Goals
+
+### Use Cases
+
+
+### Requirements
+
+* Should be able to watch v1 objects in a backward compatible way.
+
+## Proposal
+
+### Do Nothing and just shift the watch object to v1
+
+We start watching v1 objects and now store v1 objects.
+
+#### Cons
+* Database will have both v1 and v1beta1 objects. Clients like cli and UI will break due to differences in the fields.
```

Comment:
In Chains, we rely on the conversion functions to convert between `v1beta1` and `v1`. Could you not do something similar?

Our plan is to watch `v1` by default. We will provide the cluster operators to configure chains to generate `v1` or `v1beta` based provenance. Internally, we will migrate all our functions to run with `v1`. There are some deprecated fields like piplineResources that strictly need `v1beta1` so our plan is to just convert the tekton object to v1beta1 for extracting the necessary information out.

Reply:
Will you convert stored v1beta1 to v1 also? Do you store yaml in provenance? Will you convert the stored one?

Reply:
At present, we just serve base64 encoded blob in a JSON field to the client without checking whether the stored object is v1 or v1beta1.

Reply:
My recommendation:

1. Results updates to reconcile `v1` object types.
2. Results API server is updated to check the type that is being served by the `Record`. If it is a v1beta1 object type, we must decode the object, convert to v1, then re-encode before sending over the wire.
3. We enhance the `tkn-results ` cli plugin with a `convert` command, that converts a `v1beta1` api object to `v1`. This could be packaged ina container image and executed as a `Job` on a cluster.